File: apps/ios/Sources/SwabCore/Vault/VaultSync.swift
/// Vault sync (FS-07 VLT-02/VLT-04): pushes the opaque encrypted blob.
/// On 409 the client re-pulls the server version and retries once —
/// single-device POC, last write wins.
import Foundation

public protocol VaultSyncApi: Sendable {
    func pushVault(blob: String, version: Int) async throws -> VaultPushResult
    func getVault() async throws -> EncryptedVaultBlob?
}

public enum VaultSyncError: Error, Equatable, Sendable {
    case conflictPersisted

    /// G3: fixed, privacy-safe code for `ErrorReporter` — mirrors
    /// `VaultError.reportCode(for:)`.
    public static func reportCode(for error: Error) -> String {
        switch error {
        case VaultSyncError.conflictPersisted: return "conflictPersisted"
        default: return "syncFailed"
        }
    }
}

public struct VaultSync: Sendable {
    private let vault: Vault
    private let api: VaultSyncApi

    public init(vault: Vault, api: VaultSyncApi) {
        self.vault = vault
        self.api = api
    }

    public func sync() async throws {
        let (blob, version) = try await vault.getEncryptedVault()
        let result = try await api.pushVault(blob: blob, version: version)
        switch result {
        case .ok(let newVersion):
            await vault.setVaultVersion(newVersion)
        case .conflict:
            let server = try await api.getVault()
            let retryVersion = (server?.version ?? version) + 1
            let retry = try await api.pushVault(blob: blob, version: retryVersion)
            guard case .ok(let finalVersion) = retry else {
                throw VaultSyncError.conflictPersisted
            }
            await vault.setVaultVersion(finalVersion)
        }
    }
}

/// VLT-10: what `SyncScheduler` replays today. ADR-001 stage 4 replaces the
/// blob with a durable outbox (VLT-10) — that type conforms here instead and
/// the triggers above it do not change.
extension VaultSync: PendingSyncWork {
    public func flushPendingWork() async throws {
        try await sync()
    }

    public func reportCode(for error: Error) -> String {
        VaultSyncError.reportCode(for: error)
    }
}

File: apps/android/app/src/main/kotlin/com/swab/android/vault/VaultSync.kt
package com.swab.android.vault

import com.swab.android.network.ApiClient
import com.swab.android.network.EncryptedVaultBlobDto
import com.swab.android.network.VaultPushResult
import com.swab.android.observability.NoopLogger
import com.swab.android.observability.SwabLogger
import com.swab.android.sync.PendingSync

/**
 * Vault sync (FS-07 VLT-02, replayed per VLT-10): pushes the opaque
 * encrypted blob. On 409
 * the client re-pulls the server version and retries once — single-device
 * POC, last write wins. Port of apps/mobile/src/vault/sync.ts.
 */
class VaultSync(
    private val vault: Vault,
    private val apiClient: ApiClient,
    private val logger: SwabLogger = NoopLogger(),
) : PendingSync {
    class ConflictPersistedException : Exception("vault sync: conflict persisted after retry")

    /**
     * The scheduler's view of this class (SUG-AND-001). It schedules
     * [PendingSync], not a vault: ADR-001 stage 4 replaces the whole-blob push
     * with a durable per-record outbox (VLT-10), and when it does, the outbox
     * becomes the [PendingSync] and every trigger keeps working untouched.
     */
    override suspend fun flush() = syncVault()

    suspend fun syncVault() {
        val local = vault.getEncryptedVault()
        val result = apiClient.pushVault(EncryptedVaultBlobDto(local.blob, local.version))
        if (result is VaultPushResult.Ok) {
            vault.setVaultVersion(result.version)
            return
        }
        // SUG-AND-012 / G3: version numbers only — never the blob (comment,
        // not just convention: see SwabLogPrivacyTest).
        logger.event(SwabLogger.Level.INFO, "vault.sync.conflict.retry", mapOf("localVersion" to local.version))
        val server = apiClient.getVault()
        val retryVersion = (server?.version ?: local.version) + 1
        val retry = apiClient.pushVault(EncryptedVaultBlobDto(local.blob, retryVersion))
        if (retry !is VaultPushResult.Ok) {
            logger.event(SwabLogger.Level.ERROR, "vault.sync.conflict.persisted", mapOf("retryVersion" to retryVersion))
            throw ConflictPersistedException()
        }
        vault.setVaultVersion(retry.version)
    }
}
